lib/encoder.js

```diff
@@ -890,6 +907,95 @@ function defineInstanceMembers() {
+    const expectedLength = buffer.length / elemLength;
+    if ((elemLength * dimensions) !== buffer.length) {
+      throw new TypeError(`Expected buffer of subtype ${subtype} with dimensions ${dimensions} to be of size ${expectedLength}, observed size ${buffer.length}`);
+    }
+    const rv = [];
+    let offset = 0;
+    for (let i = 0; i < dimensions; i++) {
+      offset = i * elemLength;
+      rv[i] = this.decode(buffer.slice(offset, offset + elemLength), subtype);
+    }
+    return new Float32Array(rv);
+  };
+
+  /**
+   * @param {CqlVector} value
+   * @param {Array} params
+   */
+  this.encodeVector = function(value, params) {
+
+    // Evaluate params to encodeVector(), returning the computed subtype
+    function evalParams() {
+
+      if (!(value instanceof Float32Array)) {
+        throw new TypeError("Driver only supports vectors of 4 byte floating point values");
+      }
+
+      // Perform client-side validation iff we were actually supplied with meaningful type info.  In practice
+      // this will only occur when using prepared statements.
+      if (params.hasOwnProperty("subtype") && params.hasOwnProperty("dimensions")) {
+
+        const subtype = params["subtype"];
+        const dimensions = params["dimensions"];
+        if (value.length !== dimensions) {
+          throw new TypeError(`Expected vector with ${dimensions} dimensions, observed size of ${value.length}`);
+        }
+        if (subtype.code !== dataTypes.float) {
+          throw new TypeError("Driver only supports vectors of 4 byte floating point values");
+        }
+        return subtype;
+      }
+
+      return { code: dataTypes.float };
+    }
+
+    if (!Encoder.isTypedArray(value)) {
+      throw new TypeError('Expected TypedArray subclass, obtained ' + util.inspect(value));
+    }
+    if (value.length === 0) {
+      throw new TypeError("Cannot encode empty array as vector");
+    }
+
+    const subtype = evalParams();
+
+    // TypedArrays are _not_ JS arrays so explicitly convert them here before trying to write them
+    // into a buffer
+    const elems = [];
+    for (const elem of value) {
+      elems.push(this.encode(elem, subtype));
+    }
+    return Buffer.concat(elems);
+  };
+
+  /**
+   * Extract the (typed) arguments from a vector type
+   *
+   * @param {String} typeName
+   * @param {String} stringToExclude Leading string indicating this is a vector type (to be excluded when eval'ing args)
+   * @param {Function} subtypeResolveFn Function used to resolve subtype type; varies depending on type naming convention
+   * @returns {Array}
+   * @internal
+   */
+  this.parseVectorTypeArgs = function(typeName, stringToExclude, subtypeResolveFn) {
+
+    const argsStartIndex = stringToExclude.length + 1;
+    const argsLength = typeName.length - (stringToExclude.length + 2);
+    const params = parseParams(typeName, argsStartIndex, argsLength);
+    if (params.length === 1 && params[0] === "") {
+      return {};
```

Comment:
when are vector params empty? please add a comment

Reply:
This actually turns out to be a thornier question than I expected.

The empty vector params case comes about when the driver has to guess the type for an object supplied for param interpolation.  If it's a Float32Array it (correctly) guesses that it's a vector but doesn't fill in the args (subtype and dimensions)... which leads to the problem.

We prolly do have enough info to fill in what's needed there.  But I don't want to have to generate a string representing vector + args only to have it re-parsed later on; it's much better to just parse it at that point and return a representation of a vector type containing objects representing subtype and dimension.  But _that_ means we'd have to change all the subsequent logic to work with that case _and_ add earlier parsing of subtypes and dimensions in cases where we aren't guessing the type based on and input param.

That entails a bit bigger refactoring of when the args are eval'd and how they're passed around.  I'm working on this now, hoping to have it wrapped up shortly enough that I can still do the release today.

Reply:
This turned out to be trickier than I'd expected.  In order to make the change I really wanted to make we'd need Encoder.encode() to parse the vector args and that just doesn't make sense on multiple levels.  Encoder.encode() shouldn't have knowledge of individual param structure for specific custom types AND everything is built around Encoeder.encode() returning objects with an "info" attribute representing the literal subtype string.  So we'd have to add logic from levels well below Encoder.encode() and change it's return type... all just to support this.

Fortunately, there's a much better way.  We have all the information we need to build a fully parameterized type string for the "info" attribute we return from Encoder.guessDataType().  I've resisted doing this kind of thing 'cause I didn't want to be in the business of building type strings (that's really the servers job) but here we are.

Once this functionality is added to guessDataType() I've confirmed that the empty param case above is no longer necessary.  The change also required re-working some of the test cases but the overall effect there seems to be pretty positive as well.